azkaban-common/src/main/java/azkaban/version/JdbcVersionSetLoader.java

```diff
@@ -0,0 +1,237 @@
+      + "version_set vset WHERE vset.id=?";
+  private static final String SELECT_VSET_FROM_MD5 = "SELECT vset.id, vset.md5, vset.json FROM "
+      + "version_set vset WHERE vset.md5=?";
+  private static final String SELECT_ALL_VSET = "SELECT vset.id, vset.md5, vset.json FROM "
+      + "version_set vset";
+  private static final String INSERT_VSET = "INSERT INTO version_set (md5, json) values (?,?)";
+  private static final String DELETE_VSET = "DELETE FROM version_set vset WHERE vset.md5=?";
+
+  private final DatabaseOperator dbOperator;
+  private final Map<String, VersionSet> md5ToVersionSet = new HashMap<>();
+  private final Map<Integer, VersionSet> idToVersionSet = new HashMap<>();
+
+  /**
+   * This constructor should be instantiated using dependency injection only to make sure that that
+   * class remains Singleton.
+   *
+   * @param dbOperator Interface to perform db related operations.
+   */
+  @Inject
+  public JdbcVersionSetLoader(DatabaseOperator dbOperator) throws IOException {
+    this.dbOperator = dbOperator;
+    for (VersionSet versionSet : fetchAllVersionSets()) {
+      this.md5ToVersionSet.put(versionSet.getVersionSetMd5Hex(), versionSet);
+      this.idToVersionSet.put(versionSet.getVersionSetId(), versionSet);
+    }
+  }
+
+  /**
+   * This method inserts the versionSetJsonString and versionSetMd5Hex to the Table version_set and
+   * get the versionSetId autogenerated post insertion.
+   */
+  @Override
+  public synchronized int insertAndGetVersionSetId(String versionSetMd5Hex,
+      String versionSetJsonString) throws IOException {
+    final SQLTransaction<Integer> insertVersionSet = transOperator -> {
+      final VersionSetHandler versionSetHandler = new VersionSetHandler();
+      // Check if the versionSet already exists for versionSetMd5Hex
+      List<VersionSet> versionSets = transOperator
+          .query(SELECT_VSET_FROM_MD5, versionSetHandler, versionSetMd5Hex);
+      if (versionSets.size() > 1) {
+        throw new SQLException(
+            "Expected only one VersionSet for versionSetMd5Hex " + versionSetMd5Hex);
+      }
+      // Insert if it doesn't
+      if (versionSets.isEmpty()) {
+        transOperator.update(INSERT_VSET, versionSetMd5Hex, versionSetJsonString);
+        return (int) transOperator.getLastInsertId();
+      }
+      return versionSets.get(0).getVersionSetId();
+    };
+    try {
+      Integer versionSetId = this.dbOperator.transaction(insertVersionSet);
+      VersionSet versionSet = new VersionSet(versionSetJsonString, versionSetMd5Hex, versionSetId);
+      this.md5ToVersionSet.put(versionSetMd5Hex, versionSet);
+      this.idToVersionSet.put(versionSetId, versionSet);
+      return versionSetId;
+    } catch (SQLException e) {
+      logger.error("Exception occurred while inserting version set and getting version id", e);
+      throw new IOException("Unable to insert and get versionSetId", e);
```

Comment:
Why IOException? Again dao specific exception can be thrown here.

Reply:
Since VersionSetLoader interface has the definition to throw IOException, it is kept along the same lines.